### PR TITLE
Sync. problem when drive space or write permission issue

### PR DESCRIPTION
**Pull request by iiLubos**
Stop syncing and write a log message when problem with drive space or write permission.
Fixes: #2382 (hope)
Fixes: https://github.com/MerginMaps/input/issues/2808

### CODE REVIEW

File: core/merginapi.cpp [hunk 1]
```diff
@@ -3178,9 +3217,16 @@ bool MerginApi::writeData( const QByteArray &data, const QString &path )
     return false;
```
Comment (tomasMizera) [inline core/merginapi.cpp:3217]: Add coreutils::log here

File: core/merginapi.cpp [hunk 2]
```diff
@@ -1693,7 +1723,15 @@ void MerginApi::finalizeProjectPullCopy( const QString &projectFullName, const Q
...
       return;
     }
-    f.write( fTmp.readAll() );
+    QByteArray data = fTmp.readAll();
+    qint64 bytesWritten = f.write( data );
+
+    // maybe problem with permissions or drive space
+    if ( data.size() != bytesWritten )
+    {
+      CoreUtils::log( "writeData ", "Failed to write content into: " + fTmp.fileName() );
```
Comment (tomasMizera) [inline core/merginapi.cpp:1732]:
```suggestion
      CoreUtils::log( "finalizeProjectPullCopy ", "Failed to write content into: " + fTmp.fileName() );
```

File: core/merginapi.cpp [hunk 3]
```diff
@@ -385,8 +385,38 @@ void MerginApi::downloadItemReplyFinished()
...
       file.close();
+
+      // maybe problem with permissions or drive space
+      if ( data.size() != bytesWritten )
+      {
+        QString serverMsg = extractServerErrorMsg( data );
+        if ( serverMsg.isEmpty() )
+        {
+          serverMsg = r->errorString();
+        }
```
Comment (tomasMizera) [inline core/merginapi.cpp:398]: There is no error from server

File: core/merginapi.cpp [hunk 4]
```diff
@@ -385,8 +385,38 @@ void MerginApi::downloadItemReplyFinished()
...
+
+        if ( transaction.firstTimeDownload )
+        {
+          Q_ASSERT( !transaction.projectDir.isEmpty() );
+          QDir( transaction.projectDir ).removeRecursively();
+        }
+
+        int httpCode = r->attribute( QNetworkRequest::HttpStatusCodeAttribute ).toInt();
+        emit networkErrorOccurred( serverMsg, QStringLiteral( "Mergin API error: writeFile" ), httpCode, projectFullName );
+
```
Comment (tomasMizera) [inline core/merginapi.cpp:415]: This is not a network error

File: core/merginapi.cpp [hunk 5]
```diff
@@ -385,8 +385,38 @@ void MerginApi::downloadItemReplyFinished()
     QFile file( tempFilePath );
     if ( file.open( QIODevice::WriteOnly ) )
     {
-      file.write( data );
+      qint64 bytesWritten = file.write( data );
       file.close();
+
+      // maybe problem with permissions or drive space
```
Comment (tomasMizera) [inline core/merginapi.cpp:391]: It would be only not enough space left

File: core/merginapi.cpp [hunk 6]
```diff
@@ -385,8 +385,38 @@ void MerginApi::downloadItemReplyFinished()
...
+        int httpCode = r->attribute( QNetworkRequest::HttpStatusCodeAttribute ).toInt();
+        emit networkErrorOccurred( serverMsg, QStringLiteral( "Mergin API error: writeFile" ), httpCode, projectFullName );
+
+        finishProjectSync( projectFullName, false );
+
+        return;
+      }
     }
     else
     {
```
Comment (tomasMizera) [inline core/merginapi.cpp:422]: We need the same fix below

File: core/merginapi.cpp [hunk 7]
```diff
@@ -1693,7 +1733,15 @@ void MerginApi::finalizeProjectPullCopy( const QString &projectFullName, const Q
...
     }
-    f.write( fTmp.readAll() );
+    QByteArray data = fTmp.readAll();
+    qint64 bytesWritten = f.write( data );
+
+    // not enough space on file system
+    if ( data.size() != bytesWritten )
+    {
+      CoreUtils::log( "pull " + projectFullName, "Failed to write content into: " + fTmp.fileName() );
+      return;
```
Comment (PeterPetrik) [inline core/merginapi.cpp:1743]: when we return here
1/ file `f` is not closed
2/ file `f` is still present in the final folder in some random bad shape

When there is problem with sync (any), we need to guarantee that 
1/ all files are in the original/pre-update/previous version 
2/ there are no file leftovers anywhere, both in temporary folders and updated folders. 
3/ we have note in diagnostics logs about the issues

File: core/merginapi.cpp [hunk 8]
```diff
@@ -3175,12 +3224,20 @@ bool MerginApi::writeData( const QByteArray &data, const QString &path )
...
 
-  file.write( data );
+  qint64 bytesWritten = file.write( data );
   file.close();
 
+  // not enough space on file system
+  if ( data.size() != bytesWritten )
+  {
+    CoreUtils::log( "writeData ", "Failed to write content into: " + path );
+    return false;
```
Comment (PeterPetrik) [inline core/merginapi.cpp:3238]: https://github.com/MerginMaps/input/blob/780622dacf156cad443caa5fa6fd0089c21e2ae9/core/merginapi.cpp#L3178 does not check the return value of this function, so the sync is not failed.

**Review by PeterPetrik — COMMENTED** [review]
We need to add a bunch of auto-test that simulates a various error conditions during sync (e.g. we can create write-only folder, we can simulate lack of disk space, etc).

**Comment by tomasMizera** [comment]
To be continued by @wonder-sk

**Pull request closed without merging**